### ISSUE TITLE
Ansible script launches on first server in servers array instead of the last one

### ISSUE THREAD

**Issue by mkieboom** [MaxBab/vagrant-examples#1]
Although the ansible script is supposed to run on the last server in the "servers" array (based on: if index == servers.size - 1), the ansible script gets launched on the first node in the servers array.

So instead of running the ansible scripts once (after all servers are deployed), currently the ansible script is being launched for each server individually.

**Comment by MaxBab** [comment]
Hello @mkieboom 

I just tested the Vagrantfile with the ansible playbook on multiple servers and the playbook executed once for all the servers as I expected.
I'm running vagrant version - Vagrant 2.2.4

Could you provide your vagrant version and the Vagrantfile you are using, please?

**Comment by mkieboom** [comment]
Hi @MaxBab,
thanks for the quick response.

$ vagrant --version
Vagrant 2.2.5

I'm using VirtualBox as the virtualization layer which cannot launch the VM's in parallel if I'm correct. Could that be the issue that it launches them sequentially?

Here is what I did (notice that it launches ansible for each virtual machine eg: server1, client1 etc)

$ git clone https://github.com/MaxBab/vagrant-examples
$ cd vagrant-examples/example3/
$ vagrant up
Bringing machine 'server1' up with 'virtualbox' provider...
Bringing machine 'client1' up with 'virtualbox' provider...
Bringing machine 'client2' up with 'virtualbox' provider...
Bringing machine 'client3' up with 'virtualbox' provider...
==> server1: Importing base box 'centos/7'...
==> server1: Matching MAC address for NAT networking...
==> server1: Checking if box 'centos/7' version '1905.1' is up to date...
==> server1: Setting the name of the VM: example3_server1_1566829251492_94851
==> server1: Clearing any previously set network interfaces...
==> server1: Preparing network interfaces based on configuration...
    server1: Adapter 1: nat
    server1: Adapter 2: hostonly
==> server1: Forwarding ports...
    server1: 22 (guest) => 2222 (host) (adapter 1)
==> server1: Running 'pre-boot' VM customizations...
==> server1: Booting VM...
==> server1: Waiting for machine to boot. This may take a few minutes...
    server1: SSH address: 127.0.0.1:2222
    server1: SSH username: vagrant
    server1: SSH auth method: private key
==> server1: Machine booted and ready!
==> server1: Checking for guest additions in VM...
    server1: No guest additions were detected on the base box for this VM! Guest
    server1: additions are required for forwarded ports, shared folders, host only
    server1: networking, and more. If SSH fails on this machine, please install
    server1: the guest additions and repackage the box to continue.
    server1:
    server1: This is not an error message; everything may continue to work properly,
    server1: in which case you may ignore this message.
==> server1: Setting hostname...
==> server1: Configuring and enabling network interfaces...
==> server1: Rsyncing folder: /home/user/vagrant-examples/example3/ => /vagrant
==> server1: Running provisioner: file...
    server1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> server1: Running provisioner: file...
    server1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> server1: Running provisioner: shell...
    server1: Running: inline script
==> server1: Running provisioner: file...
    server1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> server1: Running provisioner: file...
    server1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> server1: Running provisioner: ansible...
Vagrant has automatically selected the compatibility mode '2.0'
according to the Ansible version installed (2.8.4).

Alternatively, the compatibility mode can be specified in your Vagrantfile:
https://www.vagrantup.com/docs/provisioning/ansible_common.html#compatibility_mode

    server1: Running ansible-playbook...
PYTHONUNBUFFERED=1 ANSIBLE_FORCE_COLOR=true ANSIBLE_HOST_KEY_CHECKING=false ANSIBLE_SSH_ARGS='-o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentityFile=/home/user/.vagrant.d/insecure_private_key -o IdentityFile=/home/user/.ssh/id_rsa -o ControlMaster=auto -o ControlPersist=60s' ansible-playbook --connection=ssh --timeout=30 --extra-vars=ansible_user\=\'vagrant\' --limit="all" --inventory-file=inventory/hosts -v playbook.yml
Using /etc/ansible/ansible.cfg as config file

PLAY [servers] *****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [server1]

TASK [Create dummy file] *******************************************************
changed: [server1] => {"changed": true, "checksum": "91b7b0b1e27bfbf7bc646946f35fa972c47c2d32", "dest": "/tmp/test_file", "gid": 1000, "group": "vagrant", "md5sum": "0b26e313ed4a7ca6904b0e9369e5b957", "mode": "0664", "owner": "vagrant", "secontext": "unconfined_u:object_r:user_home_t:s0", "size": 19, "src": "/home/vagrant/.ansible/tmp/ansible-tmp-1566829287.47-273553131585814/source", "state": "file", "uid": 1000}

TASK [Read test file content] **************************************************
ok: [server1] => {"changed": false, "content": "VGhpcyBpcyBhIHRlc3QgZmlsZQ==", "encoding": "base64", "source": "/tmp/test_file"}

TASK [Print test file content] *************************************************
ok: [server1] => {
    "msg": "This is a test file"
}

PLAY RECAP *********************************************************************
server1                    : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

==> client1: Importing base box 'ubuntu/xenial64'...
==> client1: Matching MAC address for NAT networking...
==> client1: Checking if box 'ubuntu/xenial64' version '20190822.0.0' is up to date...
==> client1: Setting the name of the VM: example3_client1_1566829295990_69003
==> client1: Fixed port collision for 22 => 2222. Now on port 2200.
==> client1: Clearing any previously set network interfaces...
==> client1: Preparing network interfaces based on configuration...
    client1: Adapter 1: nat
    client1: Adapter 2: hostonly
==> client1: Forwarding ports...
    client1: 22 (guest) => 2200 (host) (adapter 1)
==> client1: Running 'pre-boot' VM customizations...
==> client1: Booting VM...
==> client1: Waiting for machine to boot. This may take a few minutes...
    client1: SSH address: 127.0.0.1:2200
    client1: SSH username: vagrant
    client1: SSH auth method: private key
==> client1: Machine booted and ready!
==> client1: Checking for guest additions in VM...
    client1: The guest additions on this VM do not match the installed version of
    client1: VirtualBox! In most cases this is fine, but in rare cases it can
    client1: prevent things such as shared folders from working properly. If you see
    client1: shared folder errors, please make sure the guest additions within the
    client1: virtual machine match the version of VirtualBox you have installed on
    client1: your host and reload your VM.
    client1:
    client1: Guest Additions Version: 5.1.38
    client1: VirtualBox Version: 6.0
==> client1: Setting hostname...
==> client1: Configuring and enabling network interfaces...
==> client1: Mounting shared folders...
    client1: /vagrant => /home/user/vagrant-examples/example3
==> client1: Running provisioner: file...
    client1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client1: Running provisioner: file...
    client1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client1: Running provisioner: shell...
    client1: Running: inline script
    client1: dpkg-preconfigure: unable to re-open stdin: No such file or directory
    client1: Selecting previously unselected package libpython2.7-minimal:amd64.
    client1: (Reading database ...
    client1: (Reading database ... 5%
    client1: (Reading database ... 10%
    client1: (Reading database ... 15%
    client1: (Reading database ... 20%
    client1: (Reading database ... 25%
    client1: (Reading database ... 30%
    client1: (Reading database ... 35%
    client1: (Reading database ... 40%
    client1: (Reading database ... 45%
    client1: (Reading database ... 50%
    client1: (Reading database ... 55%
    client1: (Reading database ... 60%
    client1: (Reading database ... 65%
    client1: (Reading database ... 70%
    client1: (Reading database ... 75%
    client1: (Reading database ... 80%
    client1: (Reading database ... 85%
    client1: (Reading database ... 90%
    client1: (Reading database ... 95%
    client1: (Reading database ... 100%
    client1: (Reading database ...
    client1: 54245 files and directories currently installed.)
    client1: Preparing to unpack .../libpython2.7-minimal_2.7.12-1ubuntu0~16.04.4_amd64.deb ...
    client1: Unpacking libpython2.7-minimal:amd64 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Selecting previously unselected package python2.7-minimal.
    client1: Preparing to unpack .../python2.7-minimal_2.7.12-1ubuntu0~16.04.4_amd64.deb ...
    client1: Unpacking python2.7-minimal (2.7.12-1ubuntu0~16.04.4) ...
    client1: Selecting previously unselected package python-minimal.
    client1: Preparing to unpack .../python-minimal_2.7.12-1~16.04_amd64.deb ...
    client1: Unpacking python-minimal (2.7.12-1~16.04) ...
    client1: Selecting previously unselected package libpython2.7-stdlib:amd64.
    client1: Preparing to unpack .../libpython2.7-stdlib_2.7.12-1ubuntu0~16.04.4_amd64.deb ...
    client1: Unpacking libpython2.7-stdlib:amd64 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Selecting previously unselected package python2.7.
    client1: Preparing to unpack .../python2.7_2.7.12-1ubuntu0~16.04.4_amd64.deb ...
    client1: Unpacking python2.7 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Selecting previously unselected package libpython-stdlib:amd64.
    client1: Preparing to unpack .../libpython-stdlib_2.7.12-1~16.04_amd64.deb ...
    client1: Unpacking libpython-stdlib:amd64 (2.7.12-1~16.04) ...
    client1: Processing triggers for man-db (2.7.5-1) ...
    client1: Processing triggers for mime-support (3.59ubuntu1) ...
    client1: Setting up libpython2.7-minimal:amd64 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Setting up python2.7-minimal (2.7.12-1ubuntu0~16.04.4) ...
    client1: Linking and byte-compiling packages for runtime python2.7...
    client1: Setting up python-minimal (2.7.12-1~16.04) ...
    client1: Selecting previously unselected package python.
    client1: (Reading database ...
    client1: (Reading database ... 5%
    client1: (Reading database ... 10%
    client1: (Reading database ... 15%
    client1: (Reading database ... 20%
    client1: (Reading database ... 25%
    client1: (Reading database ... 30%
    client1: (Reading database ... 35%
    client1: (Reading database ... 40%
    client1: (Reading database ... 45%
    client1: (Reading database ... 50%
    client1: (Reading database ... 55%
    client1: (Reading database ... 60%
    client1: (Reading database ... 65%
    client1: (Reading database ... 70%
    client1: (Reading database ... 75%
    client1: (Reading database ... 80%
    client1: (Reading database ... 85%
    client1: (Reading database ... 90%
    client1: (Reading database ... 95%
    client1: (Reading database ... 100%
    client1: (Reading database ...
    client1: 54991 files and directories currently installed.)
    client1: Preparing to unpack .../python_2.7.12-1~16.04_amd64.deb ...
    client1: Unpacking python (2.7.12-1~16.04) ...
    client1: Processing triggers for man-db (2.7.5-1) ...
    client1: Setting up libpython2.7-stdlib:amd64 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Setting up python2.7 (2.7.12-1ubuntu0~16.04.4) ...
    client1: Setting up libpython-stdlib:amd64 (2.7.12-1~16.04) ...
    client1: Setting up python (2.7.12-1~16.04) ...
==> client1: Running provisioner: file...
    client1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client1: Running provisioner: file...
    client1: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client1: Running provisioner: ansible...
Vagrant has automatically selected the compatibility mode '2.0'
according to the Ansible version installed (2.8.4).

Alternatively, the compatibility mode can be specified in your Vagrantfile:
https://www.vagrantup.com/docs/provisioning/ansible_common.html#compatibility_mode

    client1: Running ansible-playbook...
PYTHONUNBUFFERED=1 ANSIBLE_FORCE_COLOR=true ANSIBLE_HOST_KEY_CHECKING=false ANSIBLE_SSH_ARGS='-o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentityFile=/home/user/.vagrant.d/insecure_private_key -o IdentityFile=/home/user/.ssh/id_rsa -o ControlMaster=auto -o ControlPersist=60s' ansible-playbook --connection=ssh --timeout=30 --extra-vars=ansible_user\=\'vagrant\' --limit="all" --inventory-file=inventory/hosts -v playbook.yml
Using /etc/ansible/ansible.cfg as config file

PLAY [servers] *****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [server1]

TASK [Create dummy file] *******************************************************
ok: [server1] => {"changed": false, "checksum": "91b7b0b1e27bfbf7bc646946f35fa972c47c2d32", "dest": "/tmp/test_file", "gid": 1000, "group": "vagrant", "mode": "0664", "owner": "vagrant", "path": "/tmp/test_file", "secontext": "unconfined_u:object_r:user_home_t:s0", "size": 19, "state": "file", "uid": 1000}

TASK [Read test file content] **************************************************
ok: [server1] => {"changed": false, "content": "VGhpcyBpcyBhIHRlc3QgZmlsZQ==", "encoding": "base64", "source": "/tmp/test_file"}

TASK [Print test file content] *************************************************
ok: [server1] => {
    "msg": "This is a test file"
}

PLAY RECAP *********************************************************************
server1                    : ok=4    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

==> client2: Importing base box 'centos/7'...
==> client2: Matching MAC address for NAT networking...
==> client2: Checking if box 'centos/7' version '1905.1' is up to date...
==> client2: Setting the name of the VM: example3_client2_1566829347879_50901
==> client2: Fixed port collision for 22 => 2222. Now on port 2201.
==> client2: Clearing any previously set network interfaces...
==> client2: Preparing network interfaces based on configuration...
    client2: Adapter 1: nat
    client2: Adapter 2: hostonly
==> client2: Forwarding ports...
    client2: 80 (guest) => 8080 (host) (adapter 1)
    client2: 22 (guest) => 2201 (host) (adapter 1)
==> client2: Running 'pre-boot' VM customizations...
==> client2: Booting VM...
==> client2: Waiting for machine to boot. This may take a few minutes...
    client2: SSH address: 127.0.0.1:2201
    client2: SSH username: vagrant
    client2: SSH auth method: private key
==> client2: Machine booted and ready!
==> client2: Checking for guest additions in VM...
    client2: No guest additions were detected on the base box for this VM! Guest
    client2: additions are required for forwarded ports, shared folders, host only
    client2: networking, and more. If SSH fails on this machine, please install
    client2: the guest additions and repackage the box to continue.
    client2:
    client2: This is not an error message; everything may continue to work properly,
    client2: in which case you may ignore this message.
==> client2: Setting hostname...
==> client2: Configuring and enabling network interfaces...
==> client2: Rsyncing folder: /home/user/vagrant-examples/example3/ => /vagrant
==> client2: Running provisioner: file...
    client2: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client2: Running provisioner: file...
    client2: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client2: Running provisioner: shell...
    client2: Running: inline script
==> client2: Running provisioner: file...
    client2: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client2: Running provisioner: file...
    client2: ~/.ssh/id_rsa.pub => ~/.ssh/authorized_keys
==> client2: Running provisioner: ansible...
Vagrant has automatically selected the compatibility mode '2.0'
according to the Ansible version installed (2.8.4).

Alternatively, the compatibility mode can be specified in your Vagrantfile:
https://www.vagrantup.com/docs/provisioning/ansible_common.html#compatibility_mode

    client2: Running ansible-playbook...
PYTHONUNBUFFERED=1 ANSIBLE_FORCE_COLOR=true ANSIBLE_HOST_KEY_CHECKING=false ANSIBLE_SSH_ARGS='-o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentityFile=/home/user/.vagrant.d/insecure_private_key -o IdentityFile=/home/user/.ssh/id_rsa -o ControlMaster=auto -o ControlPersist=60s' ansible-playbook --connection=ssh --timeout=30 --extra-vars=ansible_user\=\'vagrant\' --limit="all" --inventory-file=inventory/hosts -v playbook.yml
Using /etc/ansible/ansible.cfg as config file

PLAY [servers] *****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [server1]

TASK [Create dummy file] *******************************************************
ok: [server1] => {"changed": false, "checksum": "91b7b0b1e27bfbf7bc646946f35fa972c47c2d32", "dest": "/tmp/test_file", "gid": 1000, "group": "vagrant", "mode": "0664", "owner": "vagrant", "path": "/tmp/test_file", "secontext": "unconfined_u:object_r:user_home_t:s0", "size": 19, "state": "file", "uid": 1000}

TASK [Read test file content] **************************************************
ok: [server1] => {"changed": false, "content": "VGhpcyBpcyBhIHRlc3QgZmlsZQ==", "encoding": "base64", "source": "/tmp/test_file"}

TASK [Print test file content] *************************************************
ok: [server1] => {
    "msg": "This is a test file"
}

PLAY RECAP *********************************************************************
server1                    : ok=4    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

**Comment by MaxBab** [comment]
@mkieboom 

I'm using the Virtualbox provisioner as well.
I updated the vagrant version to 2.2.5 and now I see the issue.

So in 2.2.4 the index count works properly and in 2.2.5 something changed.
I will dig into this and will update the issue with my findings.

**Comment by mkieboom** [comment]
Great @MaxBab. Thanks a lot, much appreciated.
I can confirm it works perfectly fine with vagrant 2.2.4.

**Comment by MaxBab** [comment]
Hi @mkieboom 

After some debugging with no results, I opened a ticket to vagrant regarding my issue [1] and guys from vagrant (big thanks to them) pointed me to my mistake within my Vagrantfile.

I referenced to the global variable "config" instead of the local guest variable "conf" in some sections. As a result, the ansible-provisioning block didn't work.
To be honest, I'm not sure why it worked in the previous version. But right now, it works both in 2.2.4 and 2.2.5 version.

I updated the Vagrantfiles within the repo, so you can pull the changes and use it.

Thank you for opening the ticket for that issue.

[1] - https://github.com/hashicorp/vagrant/issues/11045

**Comment by MaxBab** [comment]
Hi @mkieboom ,

The issue has been resolved. The code has been updated.
I'm closing the issue.
If you have any other issues, feel free to open a new issue.

Thanks.
